modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+}
+
+/*
+  Returns a new record of type `ipAddr` prvoided `host` and `port`.
+  The family type is assumed based on `host` which is a standard address.
+  this function is equivalent to the following code:
+
+  .. code-block:: Chapel
+
+    ipAddr.ipv4(IPv4LocalHost, 8111)
+
+  :arg host: standard ipv4 address.
+  :type host: `ipv4Addr`
+  :arg port: network address's port.
+  :type port: `uint(16)`
+  :return: address instance.
+  :rtype: `ipAddr`
+  :throws SystemError: Upon incompatible `host`, `port` or `family`
+*/
+proc type ipAddr.ipv4(host: ipv4Addr, port: uint(16) = 8000): ipAddr throws {
+  var addressStorage = new sys_sockaddr_t();
+  addressStorage.set(host, port);
+  return new ipAddr(addressStorage);
+}
+
+/*
+  Returns a new record of type `ipAddr` prvoided `host` and `port`.
+  The family type is assumed based on `host` which is a standard address.
+  this function is equivalent to the following code:
+
+  .. code-block:: Chapel
+
+    ipAddr.ipv6(IPv6LocalHost, 8111)
+
+  :arg host: standard ipv6 address.
+  :type host: `ipv6Addr`
+  :arg port: network address's port.
+  :type port: `uint(16)`
+  :return: address instance.
+  :rtype: `ipAddr`
+  :throws SystemError: Upon incompatible `host`, `port` or `family`
+*/
+proc type ipAddr.ipv6(host: ipv6Addr, port: uint(16) = 8000): ipAddr throws {
+  var addressStorage = new sys_sockaddr_t();
+  addressStorage.set(host, port);
+  return new ipAddr(addressStorage);
+}
+
+/*
+  Returns the family type of address.
+  :return: family type of address
+  :rtype: :type:`IPFamily`
+*/
+proc ipAddr.family throws {
+  return _addressStorage.family:IPFamily;
+}
+
+/*
+  Returns the host address.
+  :return: host address
+  :rtype: `string`
+*/
+proc ipAddr.host throws {
+  return _addressStorage.numericHost();
+}
+
+/*
+  Returns the `port` stored in record.
+  :return: Returns numeric port.
+  :rtype: `uint(16)`
+*/
+proc ipAddr.port throws {
+  return _addressStorage.port();
+}
+
+/* compare ipAddr */
+inline operator !=(in lhs: ipAddr,in rhs: ipAddr) throws {
+  return lhs.family != rhs.family || lhs.host != rhs.host || lhs.port != rhs.port;
+}
+
+/* write ipAddr */
+proc ipAddr.writeThis(f) throws {
+  f.write("(","family:",this.family,",host:",this.host,",port:",this.port,")");
+}
+
+pragma "no doc"
+private extern proc qio_get_fd(fl:qio_file_ptr_t, ref fd:c_int):syserr;
+
+/* Type TCPConn returned from :proc:`connect` */
+type tcpConn = file;
+
+/*
+  Returns the file descriptor associated with socket
+  :return: Returns file descriptor.
+  :rtype: `int(32)`
+*/
+proc tcpConn.socketFd throws {
+  var tempfd:c_int;
+  var err:syserr = ENOERR;
+  on this.home {
+    err = qio_get_fd(this._file_internal, tempfd);
+  }
+  if err then try ioerror(err, "in tcpConn.socketFd");
+  return tempfd;
+}
+
+/*
+  Returns the address of remote socket connection
+  :return: Returns remote address.
+  :rtype: `ipAddr`
+*/
+proc tcpConn.addr throws {
+  return getpeername(this.socketFd);
```

Comment:
If the `tcpConn` is on a different locale, this function will fail, because the file descriptor will not be know to the system on the other locale. I think you need to include an `on` statement here.

Reply:
compiler won't allow me to have return statement inside of an `on` statement how should I handle that

Reply:
You have to create a temporary variable, e.g.

``` chpl
proc myproc() {
  var tmp: int;
  on somewhere {
    tmp = computeSomething();
  }
  return tmp;
}
```